plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/Maven2MetadataMerger.java

```diff
@@ -0,0 +1,352 @@
+    if (rv.getSnapshot() != null) {
+      if (lv.getSnapshot() == null) {
+        lv.setSnapshot(rv.getSnapshot());
+      }
+      else {
+        final long lts = ts(lv.getSnapshot().getTimestamp());
+        final long rts = ts(rv.getSnapshot().getTimestamp());
+        if (rts > lts) {
+          lv.setSnapshot(rv.getSnapshot());
+        }
+      }
+    }
+
+    // snapshotVersions: add ext+classifier combos, if not exist, or are older version
+    nullElementFilter(lv.getSnapshotVersions());
+    nullElementFilter(rv.getSnapshotVersions());
+    for (SnapshotVersion snapshotVersion : rv.getSnapshotVersions()) {
+      boolean found = false;
+      for (SnapshotVersion preExisting : lv.getSnapshotVersions()) {
+        if (Objects.equals(snapshotVersion.getExtension(), preExisting.getExtension())
+            && Objects.equals(nullOrEmptyStringFilter(snapshotVersion.getClassifier()),
+            nullOrEmptyStringFilter(preExisting.getClassifier()))) {
+          found = true;
+          if (version(snapshotVersion.getVersion()).compareTo(version(preExisting.getVersion())) > 0) {
+            preExisting.setClassifier(nullOrEmptyStringFilter(snapshotVersion.getClassifier()));
+            preExisting.setVersion(snapshotVersion.getVersion());
+            preExisting.setUpdated(snapshotVersion.getUpdated());
+          }
+          break;
+        }
+      }
+      if (!found) {
+        final SnapshotVersion newSnapshotVersion = new SnapshotVersion();
+        newSnapshotVersion.setExtension(snapshotVersion.getExtension());
+        newSnapshotVersion.setClassifier(snapshotVersion.getClassifier());
+        newSnapshotVersion.setVersion(snapshotVersion.getVersion());
+        newSnapshotVersion.setUpdated(snapshotVersion.getUpdated());
+        lv.getSnapshotVersions().add(newSnapshotVersion);
+      }
+    }
+  }
+
+  /**
+   * Returns last element of sorted string version list, the "latest".
+   */
+  private String getLatestVersion(final List<String> sortedVersions) {
+    return sortedVersions.get(sortedVersions.size() - 1);
+  }
+
+  /**
+   * Returns last non-snapshot element of sorted string version list, the "release". Returns {@code null} if no such
+   * element found.
+   */
+  private String getReleaseVersion(final List<String> sortedVersions) {
+    for (int i = sortedVersions.size() - 1; i >= 0; i--) {
+      if (!sortedVersions.get(i).endsWith("SNAPSHOT")) {
+        return sortedVersions.get(i);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * If string Null-Or-Empty returns empty string, otherwise the string.
+   */
+  @Nonnull
+  private String nullOrEmptyStringFilter(final String str) {
+    if (Strings.isNullOrEmpty(str)) {
+      return "";
+    }
+    return str.trim();
+  }
+
+  /**
+   * Filters a list for {@code null} (and "null" string) elements, removes them.
+   */
+  private <T> void nullElementFilter(final List<T> list) {
+    for (Iterator<T> i = list.iterator(); i.hasNext(); ) {
```

Comment:
minor but you could also use Guava to strip nulls, although it does return a view instead of modifying the original collection:
```
return Collections2.filter(list, Predicates.notNull())
```